accord-core/src/test/java/accord/local/cfk/CommandsForKeyTest.java

```diff
@@ -0,0 +1,1081 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package accord.local.cfk;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.EnumMap;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.concurrent.Callable;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import org.junit.jupiter.api.Test;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import accord.api.Agent;
+import accord.api.Data;
+import accord.api.DataStore;
+import accord.api.Key;
+import accord.api.ProgressLog;
+import accord.api.Query;
+import accord.api.Read;
+import accord.api.Result;
+import accord.api.RoutingKey;
+import accord.api.Update;
+import accord.impl.IntKey;
+import accord.local.Command;
+import accord.local.Command.AbstractCommand;
+import accord.local.CommandStore;
+import accord.local.CommandStores;
+import accord.local.CommonAttributes;
+import accord.local.Listeners;
+import accord.local.Node;
+import accord.local.NodeTimeService;
+import accord.local.PreLoadContext;
+import accord.local.SafeCommand;
+import accord.local.SafeCommandStore;
+import accord.local.SaveStatus;
+import accord.local.Status;
+import accord.local.cfk.CommandsForKey.TxnInfo;
+import accord.primitives.Ballot;
+import accord.primitives.Deps;
+import accord.primitives.EpochSupplier;
+import accord.primitives.FullRoute;
+import accord.primitives.Keys;
+import accord.primitives.Participants;
+import accord.primitives.Range;
+import accord.primitives.Ranges;
+import accord.primitives.Routable.Domain;
+import accord.primitives.Route;
+import accord.primitives.Seekable;
+import accord.primitives.Seekables;
+import accord.primitives.Timestamp;
+import accord.primitives.Txn;
+import accord.primitives.TxnId;
+import accord.primitives.Unseekables;
+import accord.primitives.Writes;
+import accord.utils.DefaultRandom;
+import accord.utils.Invariants;
+import accord.utils.RandomSource;
+import accord.utils.async.AsyncChain;
+import accord.utils.async.AsyncResults;
+
+import static accord.local.Command.NotDefined.notDefined;
+import static accord.local.Status.Durability.NotDurable;
+
+// TODO (expected): test setting redundant before
+// TODO (expected): test ballot updates
+// TODO (expected): test transition to Erased
+// TODO (expected): ensure execution is not too early
+// TODO (expected): validate mapReduce
+// TODO (expected): insert linearizability violations and detect them
+public class CommandsForKeyTest
```

Comment:
this class doesn't have any asserts... so if bad state happens its allowed...

To prove this point to myself I added the following regression....

```
static CommandsForKey Pruning.maybePrune(CommandsForKey cfk, int pruneInterval, long minHlcDelta)
    {
            // prune everything!
            if (cfk.size() == 0)
                return cfk;
            return pruneBefore(cfk, cfk.byId[0], 0);
```

I ran `testMany` and its green... we do a ton of work and spend a ton of time, but we don't validate the state of the CFK...

Reply:
We do validate quite a lot actually, though most of it depends on `checkIntegrity` in CFK, which requires command line parameters to be specified. We also verify that all of the transactions execute. What we don't currently check is that they don't execute too early, so I think this particular corruption would go undetected anyway - since an almost empty CFK is perfectly valid in itself.

Reply:
I had a TODO to detect early execution, so I just added this and that change now fails the test

Reply:
Oh, hold on. This wasn't pruning. We do fail even without updating the test if we do 

        int len = cfk.byId.length;
        return pruneBefore(cfk, cfk.byId[len - 1], len - 1);